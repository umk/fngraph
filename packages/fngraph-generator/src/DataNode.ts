import { DeclarationID } from '@fngraph/data'

import Getter from './Getter'
import PropertiesGetter from './PropertiesGetter'

type DataNode = {
  /**
   * Declarations which values the node depends on.
   */
  incoming: Array<DeclarationID>
  /**
   * Declarations which values are produced by the node.
   */
  outgoing: Array<DeclarationID>
  /**
   * A function that actually produces declarations based on
   * combinations of incoming declarations.
   */
  getter: Getter
  /**
   * Indicates whether node output is inverted. The inverted
   * nodes cannot have outgoing declarations as they just
   * serve as a predicate for incoming declarations.
   */
  invert: boolean
  /**
   * The priority the node is invoked. Higher number means the
   * node is invoked later.
   */
  priority: number
  /**
   * A function that produces a collection of properties to be
   * collected from the getter output based on other nodes inputs.
   */
  getProperties: PropertiesGetter
}

export default DataNode
