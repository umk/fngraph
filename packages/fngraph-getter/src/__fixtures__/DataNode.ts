import { DataNode } from '@fngraph/generator'

export function createDataNode(overrides?: Partial<DataNode>): DataNode {
  return {
    incoming: [],
    outgoing: [],
    getter: jest.fn(),
    priority: 0,
    invert: false,
    getProperties: () => [],
    ...overrides,
  }
}
