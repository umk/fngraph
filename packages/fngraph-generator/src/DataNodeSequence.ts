import { DeclarationID } from '@fngraph/data'

import DataNode from './DataNode'
import PriorityQueue from './PriorityQueue'

class DataNodeSequence<N extends DataNode = DataNode> {
  private constructor(
    public readonly nodes: Array<N>,
    /**
     * @internal
     */
    public readonly incomings: Map<DeclarationID, Array<N>>,
  ) {}
  static create<N extends DataNode>(nodes: Array<N>) {
    const incomings = getIncomingNodes(nodes)
    const n = getOrderedNodes(nodes, incomings)
    return new DataNodeSequence<N>(n, incomings)
  }
}

function getIncomingNodes<N extends DataNode>(nodes: Array<N>) {
  return nodes.reduce((prev, cur) => {
    for (const incoming of cur.incoming) {
      let current = prev.get(incoming)
      if (!current) prev.set(incoming, (current = []))
      current.push(cur)
    }
    return prev
  }, new Map<DeclarationID, Array<N>>())
}

function getOrderedNodes<N extends DataNode>(
  nodes: Array<N>,
  incomings: Map<DeclarationID, Array<N>>,
): Array<N> {
  const indexes = new Map(nodes.map((n) => [n, 0]))
  const pending = new PriorityQueue<N>((a, b) => a.priority - b.priority)
  nodes.forEach((n) => pending.enqueue(n))
  for (;;) {
    const current = pending.dequeue()
    if (!current) break
    // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
    const i = indexes.get(current)!
    if (i === nodes.length) {
      throw new Error('The nodes have recursive dependency')
    }
    for (const outgoing of current.outgoing) {
      const n = incomings.get(outgoing)
      if (n) {
        for (const node of n) {
          // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
          if (i >= indexes.get(node)!) {
            indexes.set(node, i + 1)
            pending.enqueue(node)
          }
        }
      }
    }
  }
  // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
  return nodes.sort((a, b) => indexes.get(b)! - indexes.get(a)!)
}

export default DataNodeSequence
