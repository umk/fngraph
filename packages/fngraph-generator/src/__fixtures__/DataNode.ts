import DataNode from '../DataNode'

export function createDataNode(overrides?: Partial<DataNode>): DataNode {
  return {
    incoming: [],
    outgoing: [],
    getter: jest.fn(),
    priority: 0,
    invert: false,
    getProperties: () => [],
    ...overrides,
  }
}
