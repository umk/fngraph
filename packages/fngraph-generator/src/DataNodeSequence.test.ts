import { DeclarationID } from '@fngraph/data'

import DataNode from './DataNode'
import DataNodeSequence from './DataNodeSequence'
import { createDataNode } from './__fixtures__'

describe('DataNodeSequence', () => {
  const declarationA = 'declaration A' as DeclarationID
  const declarationB = 'declaration B' as DeclarationID

  it('throws an error if nodes have recursive dependencies', () => {
    const nodeA: DataNode = createDataNode({ incoming: [declarationA], outgoing: [declarationB] })
    const nodeB: DataNode = createDataNode({ incoming: [declarationB], outgoing: [declarationA] })

    expect(() => DataNodeSequence.create([nodeA, nodeB])).toThrowErrorMatchingSnapshot()
  })

  it('throws an error nodes has a dependency on itself', () => {
    const nodeA: DataNode = createDataNode({ incoming: [declarationA], outgoing: [declarationA] })

    expect(() => DataNodeSequence.create([nodeA])).toThrowErrorMatchingSnapshot()
  })
})
