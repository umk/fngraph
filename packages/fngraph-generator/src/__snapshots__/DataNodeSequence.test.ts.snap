// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DataNodeSequence throws an error if nodes have recursive dependencies 1`] = `"The nodes have recursive dependency"`;

exports[`DataNodeSequence throws an error nodes has a dependency on itself 1`] = `"The nodes have recursive dependency"`;
